src/Http/Routing/src/Matching/HostMatcherPolicy.cs

```diff
@@ -0,0 +1,330 @@
+                    if (!edges.ContainsKey(contentType))
+                    {
+                        edges.Add(contentType, new List<Endpoint>());
+                    }
+                }
+            }
+
+            // Now in a second loop, add endpoints to these lists. We've enumerated all of
+            // the states, so we want to see which states this endpoint matches.
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var endpoint = endpoints[i];
+
+                var endpointKeys = endpoint.Metadata.GetMetadata<IHostMetadata>()?.Hosts.Select(h => CreateEdgeKey(h)).ToArray() ?? Array.Empty<EdgeKey>();
+                if (endpointKeys.Length == 0)
+                {
+                    // OK this means that this endpoint matches *all* hosts.
+                    // So, loop and add it to all states.
+                    foreach (var kvp in edges)
+                    {
+                        kvp.Value.Add(endpoint);
+                    }
+                }
+                else
+                {
+                    // OK this endpoint matches specific hosts
+                    foreach (var kvp in edges)
+                    {
+                        // The edgeKey maps to a possible request header value
+                        var edgeKey = kvp.Key;
+
+                        for (var j = 0; j < endpointKeys.Length; j++)
+                        {
+                            var endpointKey = endpointKeys[j];
+
+                            if (edgeKey.Equals(endpointKey))
+                            {
+                                kvp.Value.Add(endpoint);
+                                break;
+                            }
+                            else if (edgeKey.HasHostWildcard && endpointKey.HasHostWildcard &&
+                                edgeKey.Port == endpointKey.Port && edgeKey.MatchHost(endpointKey.Host))
+                            {
+                                kvp.Value.Add(endpoint);
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return edges
+                .Select(kvp => new PolicyNodeEdge(kvp.Key, kvp.Value))
+                .ToArray();
+        }
+
+        public PolicyJumpTable BuildJumpTable(int exitDestination, IReadOnlyList<PolicyJumpTableEdge> edges)
+        {
+            if (edges == null)
+            {
+                throw new ArgumentNullException(nameof(edges));
+            }
+
+            // Since our 'edges' can have wildcards, we do a sort based on how wildcard-ey they
+            // are then then execute them in linear order.
+            var ordered = edges
+                .Select(e => (host: (EdgeKey)e.State, destination: e.Destination))
+                .OrderBy(e => GetScore(e.host))
+                .ToArray();
+
+            return new HostPolicyJumpTable(exitDestination, ordered);
+        }
```

Comment:
I didn't review this in incredible detail since you've written 3/4 of this before. LMK if this one turned out to be substantially different than others.